### PR TITLE
[HEAP-28592] improve error in looker

### PR DESCRIPTION
**Pull request by yiyangli-heap**
- include the form validation in async form, the method is called when the configuration is saved;
- don't throw an error in onFields and onRow, instead, put error to errors. if we want to skip a row, don't push error to errors;
- if errors is not empty, return the failure response
- for every info, error log, try to include envId and webhookId. The webhookId can help us categorize logs for a job.

HEAP-28592

### CODE REVIEW

File: src/actions/heap/heap.ts [hunk 1]
```diff
@@ -136,25 +150,27 @@ export class HeapAction extends Hub.Action {
                 `Processed ${rowsProcessed} rows in ${
                   rowsProcessed / HeapAction.ROWS_PER_BATCH
                 } batch requests for envId ${envId}.`,
-              )
+              logTag)
             }
           }
         } catch (err) {
-          errors.push(err)
+          errors.push(new Error("Encountered an error onRow, error: " + getErrorMessage(err)))
```
Comment (yiyangli-heap) [inline src/actions/heap/heap.ts]: I want to create an InternalError here, and filter them out whent he final response is generated.

File: src/actions/heap/heap.ts [hunk 2]
```diff
@@ -165,29 +181,46 @@ export class HeapAction extends Hub.Action {
...
+      // (arbitrary limit)
+      const errorsToDisplay = errors.slice(0, HeapAction.DISPLAY_ERROR_COUNT)
+      // tell how many errors there were in total since we're only displaying the first N
+      const errorDesc = `Heap action for envId ${envId} failed with ${errors.length} errors ` +
+        (errors.length > HeapAction.DISPLAY_ERROR_COUNT ? `(displaying first ${HeapAction.DISPLAY_ERROR_COUNT})` : "")
+      winston.error(errorDesc, logTag)
+      // log first N errors
+      errorsToDisplay.forEach((err) => winston.error(`envId ${envId} error: ${err.message}`, logTag))
+      // concat first N errors into a signle errorMsg to return to the looker action hub.
+      const errorMsg = errorsToDisplay.map((err) => err.message).join(", ")
```
Comment (yiyangli-heap) [inline src/actions/heap/heap.ts:224]: filter out InternalError in the next PR. I want to deploy this version and see the response firstly, see if it's improved.

File: src/actions/heap/heap.ts [hunk 3]
```diff
@@ -267,77 +307,112 @@ export class HeapAction extends Hub.Action {
...
     // each of the following could potentially be used as keys of the field,
     // so we check each one until we find one that exists.
     const propertiesToCheck = [field.name, field.label_short, field.field_group_label]
     const fieldName = propertiesToCheck.find((f) => !!f && row.hasOwnProperty(f))
     if (fieldName) {
-      return {
+     const value = row[fieldName].value != null ? row[fieldName].value.toString() : ""
+     return {
         key: fieldName,
-        value: row[fieldName].value,
```
Comment (yiyangli-heap) [inline src/actions/heap/heap.ts:301]: I think this is why the error is 500, fieldName exists, but value is undefined. and next, value.toString is called.

File: src/actions/heap/heap.ts [hunk 4]
```diff
@@ -58,19 +70,13 @@ export class HeapAction extends Hub.Action {
   supportedActionTypes = [Hub.ActionType.Query]
   supportedPropertyTypes = [HeapPropertyTypes.User, HeapPropertyTypes.Account]
   usesStreaming = true
-  supportedFormats = (request: Hub.ActionRequest) => {
```
Comment (colbysadams) [inline src/actions/heap/heap.ts:61]: why was this removed?

File: src/actions/heap/heap.ts [hunk 5]
```diff
@@ -136,25 +150,27 @@ export class HeapAction extends Hub.Action {
...
             }
           }
         } catch (err) {
-          errors.push(err)
+          errors.push(new Error("Encountered an error onRow, error: " + getErrorMessage(err)))
         }
       },
     })
 
-    if (requestBatch.length > 0) {
```
Comment (colbysadams) [inline src/actions/heap/heap.ts:148]: why was this removed?

File: src/actions/heap/heap.ts [hunk 6]
```diff
@@ -136,25 +150,27 @@ export class HeapAction extends Hub.Action {
...
     } catch (err) {
-      errors.push(err)
+      errors.push(new Error("Encountered an error in execute, error: " + getErrorMessage(err)))
+    }
+
+    if (requestBatch.length > 0) {
+      const length = requestBatch.length
+      winston.info(`Loading the remaining ${length} rows of data to heap`, logTag)
+      await this.sendRequest(requestBatch, envId, requestUrl, heapField, errors)
     }
```
Comment (colbysadams) [inline src/actions/heap/heap.ts]: if we call this after the above try statement, I don't think that this error won't be caught in the `await Promise.all(requestPromises)`

File: src/actions/heap/heap.ts (hunk 3 above)
Comment (colbysadams) [reply src/actions/heap/heap.ts:301]: value.toString is only called after we check if the returned value was falsy... rather than attaching the error to your returned value here, I think we should throw an error when the value is undefined and then handle that error in the calling function.

File: src/actions/heap/heap.ts [hunk 7]
```diff
@@ -50,6 +61,7 @@ export class HeapAction extends Hub.Action {
   static LOG_PROGRESS_STEP = 10000
   static DISPLAY_ERROR_COUNT = 250
 
+  emptyHeapEntity = { heapFieldValue: undefined, properties: {} }
```
Comment (colbysadams) [inline src/actions/heap/heap.ts]: can we make this static?

File: src/actions/heap/heap.ts [hunk 8]
```diff
@@ -216,38 +249,45 @@ export class HeapAction extends Hub.Action {
...
-      request.formParams.heap_field.length === 0
+      !formParams.heap_field ||
+      formParams.heap_field.length === 0
     ) {
       return new Error("Column mapping to a Heap field must be provided.")
     }
   }
 
-  private extractHeapFieldByLabel(fields: Hub.Field[], heapFieldLabel: string): Hub.Field {
+  private extractHeapFieldByLabel(fields: Hub.Field[], heapFieldLabel: string, errors: Error[]): Hub.Field | undefined {
```
Comment (colbysadams) [inline src/actions/heap/heap.ts]: instead of passing in an errors array, could we instead update the calling code to catch this error when it is thrown and take the appropriate action there?

File: src/actions/heap/heap.ts [hunk 9]
```diff
@@ -386,8 +461,8 @@ export class HeapAction extends Hub.Action {
         })
         .promise()
     } catch (err) {
-      errors.push(err)
-    }
+      errors.push(new Error("Encountered an error in sending request to heap, error: " + getErrorMessage(err)))
```
Comment (colbysadams) [inline src/actions/heap/heap.ts]: Can we make this a debug log?

**Review by colbysadams — CHANGES_REQUESTED** [review]

File: src/actions/heap/heap.ts (hunk 4 above)
Comment (yiyangli-heap) [reply src/actions/heap/heap.ts:61]: I don't think it's used. I checked other actions. 

https://github.com/looker/actions/blob/e2d0d8c0261bc06d48b2a98e4327a80dcbc5ef50/src/actions/google/drive/google_drive.ts#L16-L21

https://github.com/heap/actions/blob/master/src/actions/slack/slack.ts#L23-L33

Should I add a `minimumSupportedLookerVersion`? Frankly, I don't find a dev doc.

Comment (colbysadams) [reply src/actions/heap/heap.ts:61]: https://docs.looker.com/sharing-and-publishing/action-hub#supported_data_formats

File: src/actions/heap/heap.ts (hunk 9 above)
Comment (yiyangli-heap) [reply src/actions/heap/heap.ts]: It will be logged in the end, when the whole thing is completed. The original test verifies the error. I want to make it an InternalError, so that the error is not included in the final response.

File: src/actions/heap/heap.ts (hunk 8 above)
Comment (yiyangli-heap) [reply src/actions/heap/heap.ts]: it's part of `onFields`, a callback when the stream is processed. I think what I could do is that, skip onRow when ~`heapFieldLabel`~ `identityField` is empty.

Comment (yiyangli-heap) [reply src/actions/heap/heap.ts]: the `errors` will be bubbled up to the final part that generates the response.

File: src/actions/heap/heap.ts (hunk 4 above)
Comment (yiyangli-heap) [reply src/actions/heap/heap.ts:61]: sounds good, I put it back.

File: src/actions/heap/heap.ts [hunk 10]
```diff
@@ -82,47 +94,62 @@ export class HeapAction extends Hub.Action {
...
+          identityField = this.extractHeapFieldByLabel(allFields, heapFieldLabel)
+          winston.debug(`envId ${envId} heapFieldName ${identityField} heapFieldLabel ${heapFieldLabel}`, logTag)
+          fieldMap = this.extractFieldMap(allFields)
+          winston.debug(`envId ${envId} fieldMap ${JSON.stringify(fieldMap)}`, logTag)
+        } catch(err) {
+          errors.push(new Error('Encountered errors in processing fields: ' + getErrorMessage(err)))
+        }
       },
       onRow: (row) => {
+        if (!identityField) {
```
Comment (colbysadams) [inline src/actions/heap/heap.ts]: why are we silently dropping rows if the identity field is not set?

Comment (yiyangli-heap) [reply src/actions/heap/heap.ts]: `identityField` is a required field for each row. If it's empty, heapFieldValue is empty, it will quit either way. choosing to quit early because `extractRequiredField` will throw an error if the value is empty.

**Review by colbysadams — APPROVED** [review]
